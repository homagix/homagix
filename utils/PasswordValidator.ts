export function isValidPassword(password: string) {
  return password.length >= 8
}
