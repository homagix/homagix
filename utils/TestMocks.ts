let cookie: unknown

export function useMockedCookie() {
  return {
    get value() {
      return cookie
    },
    set value(val: unknown) {
      cookie = val
    },
  }
}
