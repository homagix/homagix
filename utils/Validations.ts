export function isValidPassword(password: string) {
  return password.length >= 8
}

export function isValidURL(url: string) {
  return url?.match(/^(https?:\/\/)?([a-zA-Z0-9-]+(\.[a-zA-Z0-9-]+)+)(:[0-9]{1,5})?(\/[^\s]*)?$/) !== null
}
