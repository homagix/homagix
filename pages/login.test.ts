import { afterEach, describe, it, expect, vi } from "vitest"
import { mockNuxtImport, registerEndpoint } from "@nuxt/test-utils/runtime"
import { flushPromises, mount } from "@vue/test-utils"
import LoginPage from "./login.vue"
import { useMockedCookie } from "~/utils/TestMocks"
import { RouterLinkStub } from "@vue/test-utils"
import { setup } from "@nuxt/test-utils"

const replace = vi.fn()
const push = vi.fn()
const setServerError = vi.fn()
const reset = vi.fn()
const fetchMock = vi.fn()

vi.mock("#app", () => ({
  useRouter: vi.fn(() => ({ replace, push })),
}))

mockNuxtImport("useMessages", () => () => ({ get: vi.fn(), setServerError, reset }))
mockNuxtImport("useCookie", () => () => useMockedCookie())
registerEndpoint("/login", { method: "POST", handler: fetchMock })

function mountComponent() {
  return mount(LoginPage, {
    global: {
      stubs: {
        RouterLink: RouterLinkStub,
      },
    },
  })
}

describe("LoginPage component", () => {
  setup({ server: false })

  afterEach(() => {
    vi.restoreAllMocks()
    vi.unstubAllGlobals()
  })

  it("renders the title and description", () => {
    const wrapper = mountComponent()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it("validates the password correctly", async () => {
    const wrapper = mountComponent()
    const passwordInput = wrapper.find("#login-password")

    await passwordInput.setValue("short")
    expect(passwordInput.classes()).not.toContain("valid")

    await passwordInput.setValue("validPassword123")
    expect(passwordInput.classes()).toContain("valid")
  })

  it.skip("sets the token and navigates to '/' after successful login", async () => {
    const token = "test-token"
    fetchMock.mockImplementation(() => ({ token }))
    mockNuxtImport("useCookie", () => () => useMockedCookie())
    mockNuxtImport("useRouter", () => vi.fn(() => ({ replace })))
    const wrapper = mountComponent()

    await wrapper.find("#login-firstname").setValue("John")
    await wrapper.find("#login-password").setValue("validPassword123")
    await wrapper.find("button").trigger("click")
    await flushPromises()

    expect(replace).toBeCalledWith("/")
    const cookie = useCookie("token")
    expect(cookie.value).toEqual(token)
  })

  it.skip("displays an error message when login fails", async () => {
    const errorMessage = "Invalid credentials"
    vi.stubGlobal("$fetch", () => new Error(errorMessage))

    const wrapper = mountComponent()

    await wrapper.find("#login-firstname").setValue("John")
    await wrapper.find("#login-password").setValue("invalidPassword")
    await wrapper.find("button").trigger("click")

    expect(wrapper.find(".error").text()).toBe("Server-Fehler")
    expect(setServerError).toHaveBeenCalledWith(new Error(errorMessage))
  })

  it("resets messages on keypress", async () => {
    const wrapper = mountComponent()
    const firstnameInput = wrapper.find("#login-firstname")

    await firstnameInput.trigger("keypress")
    expect(reset).toHaveBeenCalled()
  })
})
