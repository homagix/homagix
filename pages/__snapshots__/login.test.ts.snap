// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginPage component > renders the title and description 1`] = `
"<h2 data-v-e8ce220f="" class="title is-4">Willkommen!</h2>
<p data-v-e8ce220f="">Du bist bereits registriert? Gib' bitte deinen Vornamen und dein Passwort an:</p>
<form data-v-e8ce220f="">
  <div data-v-e8ce220f="" class="fields"><label data-v-e8ce220f="" for="login-firstname"> Vorname </label><input data-v-e8ce220f="" id="login-firstname" type="text" placeholder="Vorname"><label data-v-e8ce220f="" for="login-password"> Passwort </label><input data-v-e8ce220f="" id="login-password" type="password" class="" placeholder="Passwort"></div>
  <div data-v-e8ce220f="" class="error"></div>
  <div data-v-e8ce220f="" class="button-list"><button data-v-e8ce220f="" type="submit" disabled="">Einloggen</button></div>
</form>
<p data-v-e8ce220f="">Du willst dich neu registrieren? Dann geht's <a data-v-e8ce220f="">hier entlang</a></p>"
`;
